x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java

```diff
@@ -303,97 +304,71 @@ public void analyzedPlan(
+            for (String clusterAlias : executionInfo.clusterAliases()) {
+                executionInfo.swapCluster(clusterAlias, (k, v) -> null);
+            }
 
-    private void preAnalyzeLookupIndices(List<TableInfo> indices, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
-        if (indices.size() > 1) {
-            // Note: JOINs on more than one index are not yet supported
-            listener.onFailure(new MappingException("More than one LOOKUP JOIN is not supported"));
-        } else if (indices.size() == 1) {
-            TableInfo tableInfo = indices.get(0);
-            TableIdentifier table = tableInfo.id();
-            // call the EsqlResolveFieldsAction (field-caps) to resolve indices and get field types
-            indexResolver.resolveAsMergedMapping(
-                table.index(),
-                Set.of("*"), // TODO: for LOOKUP JOIN, this currently declares all lookup index fields relevant and might fetch too many.
-                null,
-                listener.map(indexResolution -> listenerResult.withLookupIndexResolution(indexResolution))
-            );
-            // TODO: Verify that the resolved index actually has indexMode: "lookup"
-        } else {
+            // here the requestFilter is set to null, performing the pre-analysis after the first step failed
+            preAnalyzeIndices(preAnalysis.indices, executionInfo, listenerResult, null, l);
+        }).<LogicalPlan>andThen((l, listenerResult) -> {
+            assert requestFilter != null : "The second analysis shouldn't take place when there is no index filter in the request";
+            LOGGER.debug("Analyzing the plan (second attempt, without filter)");
+            LogicalPlan plan;
             try {
-                // No lookup indices specified
-                listener.onResponse(
-                    new ListenerResult(
-                        listenerResult.indices,
-                        IndexResolution.invalid("[none specified]"),
-                        listenerResult.enrichResolution,
-                        listenerResult.fieldNames
-                    )
-                );
-            } catch (Exception ex) {
-                listener.onFailure(ex);
+                plan = analyzeAction.apply(listenerResult);
+            } catch (Exception e) {
+                l.onFailure(e);
+                return;
             }
-        }
+            LOGGER.debug("Analyzed plan (second attempt, without filter):\n{}", plan);
+            l.onResponse(plan);
+        }).addListener(logicalPlanListener);
+    }
+
+    private void preAnalyzeLookupIndices(TableInfo tableInfo, ListenerResult listenerResult, ActionListener<ListenerResult> listener) {
+        TableIdentifier table = tableInfo.id();
+        Set<String> fieldNames = listenerResult.wildcardJoinIndices().contains(table.index())
+            ? IndexResolver.ALL_FIELDS
```

Comment:
@alex-spies 

> Maybe we should add a TODO comment here to remind us of the optimization potential.

Imo, TODOs are way less visible and we act way less on them compared to a **github issue**.

> Because we have only one set of fieldNames in the listenerResult, queries like FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP f1, f2, f3 either have to ask for fields field, f1, f2, f3 from both indices (slightly wasteful), or only for field from idx (correct) and * from lu_idx (quite wasteful).

**Be careful here**, we risk eliminating some edge cases. `fieldNames` should cover as many fields as possible (VS. using `*`), not as few as possible.

> FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP f1, f2, f3 either have to ask for fields `field, f1, f2, f3` from both indices (slightly wasteful)

This is not wasteful, this is the bare minimum. **This list cannot be smaller than this**. Whatever the user used in the query we need to ask field_caps to establish if the user didn't have a typo or wanted to use incompatible types fields.

> We ask for more fields than we need, and we can determine this already.

Nop. As I mentioned above, this is not a case of having the smallest list possible based on LogicalPlan tree analysis and optimization. The analysis and optimization happens after `_field_caps` call. You cannot establish which fields you need and which you don't **before** the field_caps call. `fieldNames` is somewhat blind to the logical meaning of a query, it is extracting the field names in a completely dumb way.

> I think it'd be optimal to have 1 set of field names per lookup index. This way, we can ask for fields f1, f2, f3 and field from lu_idx, and just for field from idx.

Same as above. It has the potential to be a dangerous over-optimization and, gut feeling-wise, I am against this.

> We perform the field caps requests one after the other instead of in parallel.

I agree, this should be investigated separately from this PR. Since we have this idea now, @craigtaverner please create a github issue so that we don't lose track of it. PR comments tends to be lost and forgotten after the PR is merged. This is less likely with a TODO, but a GH issue is the best reminder imo.

> When the same lookup join index is used in 2 lookup joins, we perform 2 field caps requests.

Indeed. Either in this PR, or in a new one this can be improved. GH issue if it will be a separate PR please.

Reply:
> This is not wasteful, this is the bare minimum. This list cannot be smaller than this.
This is about the query `FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP f1, f2, f3`.
The list of fields required from `idx` can be smaller; we can be certain that it is just `field`. It cannot be smaller for `lu_idx`.

In particular, for queries like `FROM idx | KEEP field | LOOKUP JOIN lu_idx ON field | KEEP *f*`, the current approach asks for `field` and `*f*` from `idx`, but it requires only `field`.

I agree that sophisticated pruning needs to be done later, in the logical plan. However, here we ask for fields and patterns, which we know are required only for lookup indices, from the main index. I recall that for certain queries a large amount of time is spent performing field caps calls, so it seemed to me that avoiding large field caps responses was important, no?

Reply:
To be clear, what I'm proposing is that we just don't mix the required fields for lookup indices into the required fields for the main index. That'd also come with a clearer separation of concerns. It should not affect, at all, the correctness of the field caps call made for the main index.

Reply:
Thanks for explaining further. This suggestion needs a completely separate issue, we need to think well about it and build it carefully. We cannot afford to miss a field from the list we send to `_field_caps`.

Reply:
@astefan and I had a quick chat offline. We concluded that it's the least risky to go with this PR's approach, but let's open a GH issue + leave a comment inside `fieldNames`. E.g. in lines 577 ff., to explain that adding the
`keepCommandReferences` to the `keepJoinReferences` ultimately makes them appear as part of the `fieldNames`, even though the field names added this way are only required from the respective lookup index, not from the main index.